File: src/tier2/cache/csharp/CacheLayer.cs
// CacheLayer.cs — Valkey TTL キャッシュ層 C# 実装（設計方針 15）
// TenantContext を key prefix にしてテナント分離を保証する
// Outbox subscribe イベントで cache invalidation を実行する
// wall-clock TTL 禁止: TTL は Stopwatch（単調増加）ベースで管理する

// 標準ライブラリのインポート
using System;
// async/await のためのインポート
using System.Threading;
using System.Threading.Tasks;
// JSON シリアライズに使用する
using System.Text.Json;
// 環境変数アクセスに使用する
using System.Collections.Generic;
// UUID 型に使用する
using System.Text;

// k1s0 tier2 cache 名前空間
namespace K1s0.Tier2.Cache
{
    // CacheKey はテナント分離付き Valkey キャッシュのキーを表す不変値オブジェクト
    public sealed record CacheKey
    {
        // TenantId: テナント識別子（テナント分離に使用する）
        public required Guid TenantId { get; init; }
        // Namespace: データ種別（"read_model" / "projection" / "aggregate"）
        public required string Namespace { get; init; }
        // Key: エンティティ識別子
        public required string Key { get; init; }

        // AsValkeyKey は Valkey に渡すフォーマット済みキー文字列を返す
        public string AsValkeyKey() =>
            // k1s0:t2:cache:{tenant_id}:{namespace}:{key} 形式でキーを生成する
            $"k1s0:t2:cache:{TenantId:D}:{Namespace}:{Key}";

        // InvalidationPattern は同テナント同 namespace 全キーにマッチするパターンを返す
        public static string InvalidationPattern(Guid tenantId, string namespaceName) =>
            // テナント + namespace の全エントリにマッチするパターンを返す
            $"k1s0:t2:cache:{tenantId:D}:{namespaceName}:*";
    }

    // CacheEntry はキャッシュエントリのメタデータを保持する
    public sealed record CacheEntry<T>
    {
        // Value: キャッシュする値
        public required T Value { get; init; }
        // CachedAtHlc: キャッシュ格納時の HLC タイムスタンプ（wall-clock TTL 禁止）
        public required string CachedAtHlc { get; init; }
        // TtlMs: TTL（ミリ秒）
        public required long TtlMs { get; init; }
    }

    // CacheLayer は Valkey TTL キャッシュ層を提供するクラス
    // テナント分離 + Outbox subscribe invalidation をサポートする
    public sealed class CacheLayer
    {
        // デフォルト TTL: 5 分（300,000 ミリ秒）
        private const long DefaultTtlMs = 300_000L;

        // _valkeyUrl: Valkey 接続 URL（環境変数 VALKEY_URL から取得する）
        private readonly string _valkeyUrl;
        // _ttlMs: 適用する TTL（ミリ秒）
        private readonly long _ttlMs;

        // コンストラクタ: Valkey URL と TTL を受け取る
        public CacheLayer(string valkeyUrl, long ttlMs = DefaultTtlMs)
        {
            // Valkey URL を設定する
            _valkeyUrl = valkeyUrl;
            // TTL を設定する
            _ttlMs = ttlMs;
        }

        // NewFromEnv は環境変数 VALKEY_URL から CacheLayer を構築するファクトリーメソッド
        public static CacheLayer NewFromEnv()
        {
            // 環境変数から Valkey URL を取得する
            var url = Environment.GetEnvironmentVariable("VALKEY_URL")
                // 未設定時のデフォルト URL
                ?? "redis://valkey.k1s0-tier2.svc:6379";
            // CacheLayer を生成して返す
            return new CacheLayer(url);
        }

        // GetAsync は指定したキャッシュキーの値を取得する
        public async Task<CacheEntry<T>?> GetAsync<T>(CacheKey cacheKey, CancellationToken ct = default)
        {
            // Valkey キーを生成する
            var vkey = cacheKey.AsValkeyKey();
            // Valkey から raw JSON を取得する
            var raw = await ValkeyGetRawAsync(vkey, ct).ConfigureAwait(false);
            // キャッシュミスの場合は null を返す
            if (raw is null) return null;
            // CacheEntry を JSON デシリアライズする
            var entry = JsonSerializer.Deserialize<CacheEntry<T>>(raw);
            // デシリアライズ失敗の場合は null を返す
            return entry;
        }

        // SetAsync は指定したキャッシュキーに値を書き込む
        public async Task SetAsync<T>(CacheKey cacheKey, T value, CancellationToken ct = default)
        {
            // Valkey キーを生成する
            var vkey = cacheKey.AsValkeyKey();
            // HLC タイムスタンプを生成する（記録目的のみ）
            var hlcTs = GenerateHlcTimestamp();
            // CacheEntry を構築する
            var entry = new CacheEntry<T>
            {
                // 値を設定する
                Value = value,
                // HLC タイムスタンプを設定する
                CachedAtHlc = hlcTs,
                // TTL を設定する
                TtlMs = _ttlMs,
            };
            // CacheEntry を JSON シリアライズする
            var json = JsonSerializer.Serialize(entry);
            // TTL を秒単位に変換する
            var ttlSecs = Math.Max(_ttlMs / 1000, 1);
            // Valkey に書き込む（SETEX コマンド）
            await ValkeySetexRawAsync(vkey, ttlSecs, json, ct).ConfigureAwait(false);
        }

        // InvalidateByOutboxAsync は Outbox subscribe イベントを受けてキャッシュを無効化する
        public async Task<long> InvalidateByOutboxAsync(Guid tenantId, string namespaceName, CancellationToken ct = default)
        {
            // 無効化パターンを生成する
            var pattern = CacheKey.InvalidationPattern(tenantId, namespaceName);
            // SCAN + UNLINK でパターンマッチするキーを削除する
            return await ValkeyScanUnlinkAsync(pattern, ct).ConfigureAwait(false);
        }

        // ValkeyGetRawAsync は Valkey GET コマンドを実行する（stub 実装）
        private Task<string?> ValkeyGetRawAsync(string key, CancellationToken ct)
        {
            // production では StackExchange.Redis の StringGetAsync を使用する
            _ = key;
            _ = ct;
            // stub としてキャッシュミスを返す
            return Task.FromResult<string?>(null);
        }

        // ValkeySetexRawAsync は Valkey SETEX コマンドを実行する（stub 実装）
        private Task ValkeySetexRawAsync(string key, long ttlSecs, string value, CancellationToken ct)
        {
            // production では StackExchange.Redis の StringSetAsync を使用する
            _ = key;
            _ = ttlSecs;
            _ = value;
            _ = ct;
            // 正常終了を返す
            return Task.CompletedTask;
        }

        // ValkeyScanUnlinkAsync は SCAN + UNLINK でパターンマッチするキーを削除する（stub 実装）
        private Task<long> ValkeyScanUnlinkAsync(string pattern, CancellationToken ct)
        {
            // production では SCAN カーソルループ + UNLINK バッチを実装する
            _ = pattern;
            _ = ct;
            // 0 件削除を返す
            return Task.FromResult(0L);
        }

        // GenerateHlcTimestamp は HLC タイムスタンプ文字列を生成する
        // wall-clock TTL 禁止規約に従い、記録目的のみに使用する
        private static string GenerateHlcTimestamp()
        {
            // DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() は記録目的のみ使用許可
            // NOTE: TTL/deadline 計算への使用は禁止（HLC ライブラリが整備されたら移行する）
            var ms = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            // HLC 形式: {ms_hex_16}-{logical_0000}-{node_0000}
            return $"{ms:x16}-0000-0000";
        }
    }
}

File: src/tier2/cache/typescript/cache_layer.ts
// cache_layer.ts — Valkey TTL キャッシュ層 TypeScript 実装（設計方針 15）
// TenantContext を key prefix にしてテナント分離を保証する
// Outbox subscribe イベントで cache invalidation を実行する
// wall-clock TTL 禁止: TTL は HLC ベースで管理する

// デフォルト TTL: 5 分（300,000 ミリ秒）
export const DEFAULT_CACHE_TTL_MS = 300_000;

// Valkey キャッシュキーのオプション
export interface CacheKeyOptions {
  // tenantId: テナント識別子（テナント分離に使用する）
  readonly tenantId: string;
  // namespace: データ種別（"read_model" / "projection" / "aggregate"）
  readonly namespace: string;
  // key: エンティティ識別子
  readonly key: string;
}

// asValkeyKey はキャッシュキーオプションから Valkey キー文字列を生成する
export function asValkeyKey(opts: CacheKeyOptions): string {
  // k1s0:t2:cache:{tenant_id}:{namespace}:{key} 形式でキーを生成する
  return `k1s0:t2:cache:${opts.tenantId}:${opts.namespace}:${opts.key}`;
}

// invalidationPattern は同テナント同 namespace の全キーにマッチするパターンを返す
export function invalidationPattern(tenantId: string, namespace: string): string {
  // テナント + namespace の全エントリにマッチするパターンを返す
  return `k1s0:t2:cache:${tenantId}:${namespace}:*`;
}

// CacheEntry はキャッシュエントリのメタデータを保持するインターフェース
export interface CacheEntry<T> {
  // value: キャッシュする値
  readonly value: T;
  // cachedAtHlc: キャッシュ格納時の HLC タイムスタンプ
  readonly cachedAtHlc: string;
  // ttlMs: TTL（ミリ秒）
  readonly ttlMs: number;
}

// HLC タイムスタンプを生成する（wall-clock TTL 禁止規約に従い記録目的のみ使用する）
function generateHlcTimestamp(): string {
  // Date.now() は記録目的のみ使用許可（TTL/deadline 計算への使用は禁止）
  // NOTE: production では @k1s0/hlc ライブラリを使用することを推奨する
  const ms = Date.now();
  // HLC 形式: {ms_hex_16}-{logical_0000}-{node_0000}
  const hex = ms.toString(16).padStart(16, '0');
  // HLC 文字列を返す
  return `${hex}-0000-0000`;
}

// ValkeyClient は Valkey（Redis 互換）クライアントのインターフェース
// production では ioredis または node-redis を使用する
export interface ValkeyClient {
  // get: キーに対応する値を取得する
  get(key: string): Promise<string | null>;
  // setex: TTL 付きでキーと値を設定する
  setex(key: string, ttlSecs: number, value: string): Promise<void>;
  // scan: パターンマッチするキーを列挙する
  scan(cursor: string, pattern: string): Promise<[string, string[]]>;
  // unlink: キーを非同期削除する（KEYS/DEL の production-safe 代替）
  unlink(keys: string[]): Promise<number>;
}

// CacheLayer は Valkey TTL キャッシュ層を提供するクラス
export class CacheLayer {
  // _valkeyClient: 注入された Valkey クライアント
  private readonly _valkeyClient: ValkeyClient;
  // _ttlMs: 適用する TTL（ミリ秒）
  private readonly _ttlMs: number;

  // constructor は Valkey クライアントと TTL を受け取る
  constructor(valkeyClient: ValkeyClient, ttlMs = DEFAULT_CACHE_TTL_MS) {
    // Valkey クライアントを設定する
    this._valkeyClient = valkeyClient;
    // TTL を設定する
    this._ttlMs = ttlMs;
  }

  // withTtl は TTL をカスタム値に設定した新しい CacheLayer を返す
  withTtl(ttlMs: number): CacheLayer {
    // 新しいインスタンスを返す
    return new CacheLayer(this._valkeyClient, ttlMs);
  }

  // get は指定したキャッシュキーの値を取得する
  async get<T>(opts: CacheKeyOptions): Promise<CacheEntry<T> | null> {
    // Valkey キーを生成する
    const vkey = asValkeyKey(opts);
    // Valkey から raw JSON を取得する
    const raw = await this._valkeyClient.get(vkey);
    // キャッシュミスの場合は null を返す
    if (raw === null) {
      // キャッシュミスを返す
      return null;
    }
    // CacheEntry を JSON パースして返す
    const entry = JSON.parse(raw) as CacheEntry<T>;
    // キャッシュエントリを返す
    return entry;
  }

  // set は指定したキャッシュキーに値を書き込む
  async set<T>(opts: CacheKeyOptions, value: T): Promise<void> {
    // Valkey キーを生成する
    const vkey = asValkeyKey(opts);
    // HLC タイムスタンプを生成する（記録目的のみ）
    const hlcTs = generateHlcTimestamp();
    // CacheEntry を構築する
    const entry: CacheEntry<T> = {
      // 値を設定する
      value,
      // HLC タイムスタンプを設定する
      cachedAtHlc: hlcTs,
      // TTL を設定する
      ttlMs: this._ttlMs,
    };
    // CacheEntry を JSON シリアライズする
    const json = JSON.stringify(entry);
    // TTL を秒単位に変換する（最低 1 秒）
    const ttlSecs = Math.max(Math.floor(this._ttlMs / 1000), 1);
    // Valkey に SETEX で書き込む
    await this._valkeyClient.setex(vkey, ttlSecs, json);
  }

  // invalidateByOutbox は Outbox subscribe イベントを受けてキャッシュを無効化する
  async invalidateByOutbox(tenantId: string, namespace: string): Promise<number> {
    // 無効化パターンを生成する
    const pattern = invalidationPattern(tenantId, namespace);
    // 削除したキー数を初期化する
    let deletedCount = 0;
    // SCAN カーソルを初期化する
    let cursor = '0';
    // SCAN ループでパターンマッチするキーを収集する
    do {
      // SCAN コマンドを実行する
      const [nextCursor, keys] = await this._valkeyClient.scan(cursor, pattern);
      // 次のカーソルを更新する
      cursor = nextCursor;
      // マッチするキーが存在する場合は UNLINK で削除する
      if (keys.length > 0) {
        // UNLINK コマンドで非同期削除する
        const count = await this._valkeyClient.unlink(keys);
        // 削除件数を加算する
        deletedCount += count;
      }
    // カーソルが '0' になるまでループする（全スキャン完了）
    } while (cursor !== '0');
    // 削除件数を返す
    return deletedCount;
  }
}
